consult.el

```diff
@@ -2149,5 +2149,87 @@ if not available use `default-directory'."
 
 (add-hook 'consult--completion-refresh-hook #'consult--icomplete-refresh)
 
+
+;;;; find, fd, locate
+(defun consult--guess-initial-dir (&optional askdir)
+  "Returns root directory for use in `consult' searching functions.
+If ASKDIR is not `nil' prompt the user for the inital directory,
+else use simple logic: first try `project', then `projectile',
+and defaults to `default-directory'."
+  (expand-file-name
+   (if askdir
+       (read-directory-name "Initial directory: ")
+     (cond
+      ((and (fboundp 'project-current) (project-current))
+       (project-root (project-current)))
+      ((and (fboundp 'projectile-project-root) (projectile-project-root))
+       (projectile-project-root))
+      (default-directory default-directory)
+      (t "~")))))
+
+
+(defun consult--find-async (cmd)
+  "Async function for `consult-find' and similar functions.
+CMD is the locate argument list."
+  (thread-first (consult--async-sink)
+    (consult--async-refresh-timer)
+    (consult--async-process cmd)
+    (consult--async-throttle)
+    (consult--async-split)))
+
+
+(defvar consult--find-cmd '("find"))
+(defvar consult--fd-cmd '("fd" "-c" "never" "-H" "-p"))
+(defvar consult--locate-cmd '("locate" "-iebr"))
+
+
+;;;###autoload
+(defun consult-find (&optional arg)
+  "Search for REGEXP with find.
+If ASKUSER is not `nil', prompt the user, else use simple logic."
+  (interactive "P")
+  (find-file
+   (let ((initialpath (consult--guess-initial-dir askuser)))
+     (consult--read
+      (concat "Find in " (propertize initialpath 'face 'consult-file) ": ")
+      (consult--find-async (lambda (input)
+                             (append consult--find-cmd (list initialpath "-ipath" input))))
+      :sort nil
+      :require-match t
+      :initial consult-async-default-split
+      :category 'file
+      :history 'file-name-history))))
+
+;;;###autoload
+(defun consult-fd (&optional askuser)
+  "Search for REGEXP with fd.
+If ASKUSER is not `nil', prompt the user, else use simple logic."
+  (interactive "P")
+  (find-file
+   (let ((initialpath (consult--guess-initial-dir askuser)))
+     (consult--read
+      (concat "FD in " (propertize initialpath 'face 'consult-file) ": ")
+      (consult--find-async (lambda (input)
+                             (append consult--fd-cmd (list input initialpath))))
+      :sort nil
+      :require-match t
+      :category 'file
+      :history 'file-name-history))))
```

Comment:
I guess it would make sense to extract (find-file (consult--read ...)) into a separate function `consult--find`. This function could then be used by `consult-find`, `consult-fd`, `consult-locate`.

Reply:
I've tried it, but it was not possible, as each function needs different arguments and `input` is passed in different order in the lambda.

``` emacs-lisp
(append consult--find-cmd (list initialpath "-ipath" input))
(append consult--fd-cmd (list input initialpath))
(append consult--locate-cmd (list input))
```

Reply:
Make it a higher order function?

Reply:
Sorry, this is above my skills. I'm not a programmer.
Anyway, wouldn't it be too convoluted for just these functions?

Reply:
Hey! I think what you mean!

This seems to work...

``` emacs-lisp
(defun consult--find (prompt func)
  "Search for REGEXP with locate."
  (find-file
   (consult--read
    prompt
    (consult--find-async func)
    :sort nil
    :require-match t
    :category 'file
    :history 'file-name-history)))

(defun consult-locate ()
  "Search for REGEXP with locate."
  (interactive)
  (consult--find "Locate: "
                 (lambda (input) (append consult--locate-cmd (list input)))))
```